mm2src/coins/utxo/utxo_common.rs

```diff
@@ -855,11 +856,76 @@ where
     Box::new(fut.boxed().compat())
 }
 
+/// Extracts pubkey from script sig
+fn pubkey_from_script_sig(script: &Script) -> Result<H264, String> {
+    match script.get_instruction(0) {
+        Some(Ok(instruction)) => match instruction.opcode {
+            Opcode::OP_PUSHBYTES_71 | Opcode::OP_PUSHBYTES_72 => match instruction.data {
+                Some(bytes) => try_s!(Signature::parse_der(&bytes[..bytes.len() - 1])),
+                None => return ERR!("No data at instruction 0 of script {:?}", script),
+            },
+            _ => return ERR!("Unexpected opcode {:?}", instruction.opcode),
+        },
+        Some(Err(e)) => return ERR!("Error {} on getting instruction 0 of script {:?}", e, script),
+        None => return ERR!("None instruction 0 of script {:?}", script),
+    };
+
+    let pubkey = match script.get_instruction(1) {
+        Some(Ok(instruction)) => match instruction.opcode {
+            Opcode::OP_PUSHBYTES_33 => match instruction.data {
+                Some(bytes) => try_s!(PublicKey::parse_slice(bytes, None)),
+                None => return ERR!("No data at instruction 1 of script {:?}", script),
+            },
+            _ => return ERR!("Unexpected opcode {:?}", instruction.opcode),
+        },
+        Some(Err(e)) => return ERR!("Error {} on getting instruction 1 of script {:?}", e, script),
+        None => return ERR!("None instruction 1 of script {:?}", script),
+    };
+
+    if script.get_instruction(2).is_some() {
+        return ERR!("Unexpected instruction at position 2 of script {:?}", script);
+    }
+    Ok(pubkey.serialize_compressed().into())
+}
+
+pub async fn is_tx_confirmed_before_block<T>(coin: &T, tx: &RpcTransaction, block_number: u64) -> Result<bool, String>
+where
+    T: AsRef<UtxoCoinFields> + Send + Sync + 'static,
+{
+    match tx.height {
+        Some(confirmed_at) => Ok(confirmed_at <= block_number),
+        // fallback to a number of confirmations
+        None => {
+            if tx.confirmations > 0 {
+                let current_block = try_s!(coin.as_ref().rpc_client.get_block_count().compat().await);
+                let confirmed_at = current_block + 1 - tx.confirmations as u64;
+                Ok(confirmed_at <= block_number)
+            } else {
+                Ok(false)
+            }
+        },
+    }
+}
+
+pub fn check_all_inputs_signed_by_pub(tx: &UtxoTx, expected_pub: &[u8]) -> Result<bool, String> {
+    for input in &tx.inputs {
```

Comment:
Just suggestion: it may be replaced by `Iterator::any`

Reply:
Yeah, it might be a bit more idiomatic, I guess :slightly_smiling_face:

Reply:
Hmm, the `pubkey_from_script_sig` can return an error, so it looks like we can't use `Iterator::any` here.

Reply:
I see :+1: